### PR TITLE
Tila 2212 - OidcProvider --> Next-auth

### PR DESCRIPTION
**Pull request by wuahi**
- Updates ui/ app:s authentication to use next-auth and new style of authentication flow
 - Updates navigation component as an example of the  styled-components: themeProvider

### CODE REVIEW

File: ui/components/LoginFragment.tsx [hunk 1]
```diff
@@ -53,53 +52,42 @@ const LoginFragment = ({
...
   const [shouldLogin, setShouldLogin] = React.useState(false);
 
   if (shouldLogin) {
-    return (
-      <RequireAuthentication>
-        <div />
-      </RequireAuthentication>
-    );
+    signIn("tunnistamo", {
+      callbackUrl: "https://local-tilavaraus.hel.fi:3000/",
```
Comment (kieferbonk) [inline ui/components/LoginFragment.tsx:64]: Proper url as a variable

File: ui/components/common/Navigation/Navigation.tsx [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+const StyledNavigation = styled(HDSNavigation)`
+  color: ${(props) => props.theme.colors.black.medium};
+  min-width: ${(props) => props.theme.breakpoints.xs};
+
+  /* .btn-logout {
+    display: flex;
+    margin-top: var(--spacing-m);
+    cursor: pointer;
+  }*/
```
Comment (kieferbonk) [inline ui/components/common/Navigation/Navigation.tsx:18]: Commented lines?

File: ui/components/common/Navigation/Navigation.tsx [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    right: 0;
+    left: unset;
+  }
+
+  @media (max-width: ${(props) => props.theme.breakpoints.xs}) {
+    position: fixed !important;
+    z-index: ${(props) => props.theme.zIndex.navigation} !important;
+  }
+
+  /* @media (max-width:${(props) => props.theme.breakpoints.m}) {
```
Comment (kieferbonk) [inline ui/components/common/Navigation/Navigation.tsx:30]: Should use mobile-first approach.

File: ui/components/common/Navigation/NavigationUserMenu/NavigationUserMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    .includes(router.pathname);
+
+  const handleSignIn = () => {
+    signIn("tunnistamo", {
+      callbackUrl: baseUrl,
+    });
+  };
+
+  const handleSignOut = () => {
+    router.push("/api/auth/logout");
```
Comment (kieferbonk) [inline ui/components/common/Navigation/NavigationUserMenu/NavigationUserMenu.tsx:114]: Value as a global constant

File: ui/components/common/Navigation/NavigationUserMenu/NavigationUserMenu.tsx [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    publicRuntimeConfig: { baseUrl },
+  } = getConfig();
+
+  const user = session.data?.user;
+  const isActive = userMenuItems
+    .map((item) => item.path)
+    .includes(router.pathname);
+
+  const handleSignIn = () => {
+    signIn("tunnistamo", {
```
Comment (kieferbonk) [inline ui/components/common/Navigation/NavigationUserMenu/NavigationUserMenu.tsx:108]: Value as a global constant

File: ui/components/common/Navigation/NavigationMenuItem.tsx [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import { Navigation as HDSNavigation } from "hds-react";
+import Link from "next/link";
+import styled from "styled-components";
+
+const NavigationMenuItem = styled(HDSNavigation.Item)`
+  white-space: nowrap !important;
```
Comment (kieferbonk) [inline ui/components/common/Navigation/NavigationMenuItem.tsx:5]: Rather use specification (`&&`) than `!important`-tag

File: ui/components/common/Navigation/Navigation.tsx [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+  const handleNavigationTitleClick = () => router.push("/");
+
+  return (
+    <StyledNavigation
+      title={t("common:applicationName")}
+      onTitleClick={handleNavigationTitleClick}
+      menuToggleAriaLabel="Menu"
+      skipTo="main"
+      skipToContentLabel="Siirry sivun pääsisältöön"
```
Comment (kieferbonk) [inline ui/components/common/Navigation/Navigation.tsx:61]: Translation needed

File: ui/components/common/Navigation/NavigationUserMenu/NavigationUserMenuUserCard.tsx [hunk 8]
```diff
@@ -0,0 +1,51 @@
+import React from "react";
+import styled from "styled-components";
+
+const StyledUserInfo = styled.div`
+  word-break: break-all;
+  padding-bottom: ${(props) => props.theme.spacing.s};
+  padding-left: ${(props) => props.theme.spacing.s};
+  /* padding: 0 0 var(--spacing-s) 36px; */
```
Comment (kieferbonk) [inline ui/components/common/Navigation/NavigationUserMenu/NavigationUserMenuUserCard.tsx:8]: Uncomment/cleanup

File: ui/components/reservation-unit/Head.tsx [hunk 9]
```diff
@@ -114,6 +115,7 @@ const Head = ({
   isReservable,
   subventionSuffix,
 }: PropsType): JSX.Element => {
+  const { asPath } = useRouter();
```
Comment (kieferbonk) [inline ui/components/reservation-unit/Head.tsx]: Cleanup?

File: ui/.env.local.example [hunk 10]
```diff
@@ -1,15 +1,36 @@
...
+# AUTH
+# ---- NOTE: SECRETS ----
+# For NEXTAUTH_SECRET, use: `openssl rand -base64 32`
+# For NEXT_PUBLIC_OIDC_AUTH_SECRET, use `openssl rand -hex 32`
+# Note: Changing a secret may invalidate existing sessions and/or verification tokens
+# -----------------------
+DISABLE_AUTH=false
+NEXTAUTH_SECRET=
+NEXT_PUBLIC_OIDC_CLIENT_ID=
+NEXT_PUBLIC_OIDC_AUTH_SECRET=
```
Comment (kieferbonk) [inline ui/.env.local.example:15]: Discrepancy with next.config

File: ui/pages/_app.tsx [hunk 11]
```diff
@@ -53,19 +49,15 @@ const MyApp = ({ Component, pageProps }: AppProps) => {
...
       <DataContextProvider>
         <TrackingWrapper>
-          <OidcProvider
-            configuration={oidcConfiguration}
-            callbackSuccessComponent={LoggingIn}
-            loadingComponent={FullscreenSpinner}
-            authenticatingComponent={FullscreenSpinner}
-            sessionLostComponent={SessionLost}
-          >
+          <SessionProvider session={pageProps.session} basePath="/api/auth">
```
Comment (kieferbonk) [inline ui/pages/_app.tsx:52]: `basePath` value as a global constant

File: ui/tsconfig.tsbuildinfo [hunk 12]
```diff
@@ -1 +1 @@
...,1],[363,1],[358,1],[360,1],[1166,1],[1169,1],[1164,1],[1168,1],[1170,1],[1167,1],[1165,1],[1160,1],[1163,1],[1162,1],[1161,1],[1525,1],[1528,1],[1527,1],[1526,1],[1414,1],[1325,1],[1327,1],[1322,1],[1321,1],[1334,1],[1326,1],[1330,1],[1331,1],[1332,1],[1320,1],[1319,1],[1323,1],[1324,1],[1328,1],[1333,1],[1329,1],[1352,1],[1359,1],[1368,1],[1353,1],[1373,1],[1362,1],[1365,1],[1360,1],[1364,1],[1366,1],[1363,1],[1361,1],[1367,1],[1358,1],[1355,1],[1357,1],[1351,1],[1354,1],[1356,1],[1376,1],[1382,1],[1375,1],[1380,1],[1379,1],[1381,1],[1377,1],[1374,1],[1378,1],[1372,1],[1371,1],[1370,1],[1369,1],[1416,1],[1417,1],[1426,1],[1418,1],[1425,1],[1424,1],[1419,1],[1421,1],[1422,1],[1423,1],[1420,1],[1404,1],[1427,1],[1403,1],[1384,1],[1383,1],[1335,1],[1349,1],[1350,1],[1336,1],[1343,1],[1342,1],[1337,1],[1345,1],[1340,1],[1341,1],[1338,1],[1339,1],[1344,1],[1348,1],[1314,1],[1346,1],[1315,1],[1316,1],[1317,1],[1318,1],[1347,1],[1397,1],[1398,1],[1396,1],[1388,1],[1387,1],[1402,1],[1399,1],[1395,1],[1389,1],[1394,1],[1392,1],[1391,1],[1390,1],[1393,1],[1385,1],[1400,1],[1386,1],[1401,1],[1241,1],[1243,1],[1283,1],[1277,1],[1245,1],[1247,1],[1244,1],[1248,1],[1246,1],[1264,1],[1278,1],[1
```
Comment (kieferbonk) [inline ui/tsconfig.tsbuildinfo:1]: Should this file be on vc?

File: ui/components/calendar/ReservationCalendarControls.tsx [hunk 13]
```diff
@@ -498,9 +498,6 @@ const ReservationCalendarControls = <T extends Record<string, unknown>>({
     <SubmitButtonWrapper>
       <LoginFragment
         isActionDisabled={!isReservable}
-        actionCallback={() => {
```
Comment (kieferbonk) [inline ui/components/calendar/ReservationCalendarControls.tsx:501]: Rollback?

Comment (wuahi) [reply ui/components/calendar/ReservationCalendarControls.tsx:501]: yeah, i was afraid of something like this. what should be here?

Comment (kieferbonk) [reply ui/components/calendar/ReservationCalendarControls.tsx:501]: Well, the original content I guess

File: ui/.env.local.example [hunk 14]
```diff
@@ -1,15 +1,31 @@
...
-#NEXT_PUBLIC_COOKIEHUB_ENABLED=true
-#NEXT_PUBLIC_MATOMO_ENABLED=true
-#NEXT_PUBLIC_HOTJAR_ENABLED=true
-#NEXT_PUBLIC_SENTRY_DSN=
-#NEXT_PUBLIC_SENTRY_ENVIRONMENT=
-#NEXT_PUBLIC_MOCK_REQUESTS=true
-#NEXT_PUBLIC_MAPBOX_TOKEN=
+# ---- NOTE: SECRETS ----
+# For NEXTAUTH_SECRET, use: `openssl rand -base64 32`
+# For OIDC_AUTH_SECRET, use `openssl rand -hex 32`
```
Comment (kieferbonk) [inline ui/.env.local.example:3]: This refers to `OIDC_CLIENT_SECRET`?

File: ui/next.config.js [hunk 15]
```diff
@@ -11,23 +11,28 @@ const nextConfig = {
...
     authEnabled: process.env.DISABLE_AUTH !== "true",
+    oidcClientId: process.env.OIDC_CLIENT_ID,
+    oidcClientSecret: process.env.OIDC_CLIENT_SECRET,
+    oidcIssuer: process.env.OIDC_URL,
+    oidcTokenUrl: process.env.OIDC_TOKEN_URL,
+    oidcAccessTokenUrl: process.env.OIDC_ACCESS_TOKEN_URL,
+    oidcScope: process.env.OIDC_SCOPE,
+    oidcCallbackUrl: process.env.OIDC_CALLBACK_URL,
+    oidcProfileApiUrl: process.env.OIDC_PROFILE_API_SCOPE,
+    oidcTilavarausApiUrl: process.env.OIDC_TILAVARAUS_API_SCOPE,
```
Comment (kieferbonk) [inline ui/next.config.js:22]: Reminder: new variables should be configured to build environments

File: ui/pages/api/auth/[...nextauth].ts [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+  const { data } = response;
+
+  if (!data) {
+    throw new Error("No api-tokens present");
+  }
+
+  const apiAccessToken: string = data[oidcTilavarausApiUrl];
+  const profileApiAccessToken: string = data[oidcProfileApiUrl];
+
+  return [apiAccessToken, profileApiAccessToken];
```
[the remark this answers is not in the document]
Comment (kieferbonk) [reply ui/pages/api/auth/[...nextauth].ts:123]: No - the user would just end up getting invalid tokens and no authenticated data.

**Review by joonatank — APPROVED** [review]

**Pull request merged**